boxes (#950)
algorand/docs: docs/get-details/dapps/smart-contracts/apps/index.md

The other arrays are used to load data from the blockchain:
* The application array is used to pass other smart contract IDs that can be used to read state for those specific contracts. 
* The assets array is used to pass a list of asset IDs that can be used to retrieve configuration and asset balance information. 
* The accounts array allows additional accounts to be passed to the contract for balance information and local state. Note that to access an account's asset balance, both the account and the asset ID must be specified in their respective arrays. Similarly, to access an account's local state for a specific application, both the account and the smart contract ID must be specified in their respective arrays.
* The boxes array defines which boxes can be manipulated in a particular call to the smart contract. The box array is an array of pairs: the first element of each pair is an integer specifying the index into the application array, and the second element is the key name of the box to be accessed.   

> These three arrays (*applications*, *assets*, and *accounts*) are limited to eight total values combined, and of those, the accounts array can have no more than four values. The values passed within these arrays can change per Application Transaction. Many opcodes that make use of these arrays take an integer parameter as an index into these arrays. The accounts and applications arrays contain the transaction sender and current application ID in the 0th position of the respective array. This shifts the contents of these two arrays by one slot. Most of the opcodes that use an index into these arrays also allow passing the actual value. For example, an address can be specified for an opcode that uses the accounts array. IDs can be specified for contracts and assets for an opcode that uses the applications or assets arrays, respectively. These opcodes will fail if the specified value does not exist in the corresponding array. The use of each of these arrays is detailed throughout this guide.

Reviewer: ```suggestion
These four arrays (*applications*, *assets*, *accounts*, and *boxes*) are limited to eight total values combined, and of those, the accounts array can have no more than four values. The values passed within these arrays can change per Application Transaction.  The opcodes that make use of these arrays take an integer parameter as an index into these arrays. 

The accounts and applications arrays contain the transaction sender and current application ID in the 0th position of the respective array. This shifts the contents of these two arrays by one slot. The opcodes that use an index into these arrays also allow passing the actual value. For example, an address can be specified for an opcode that uses the accounts array. IDs can be specified for contracts and assets for an opcode that uses the applications or assets arrays, respectively. These opcodes will fail if the specified value does not exist in the corresponding array. The use of each of these arrays is detailed throughout this guide.

!! Note
  The boxes array cannot be accessed directly with an opcode like the other arrays.
```

Since TEAL 6, all created assets and apps are available to be accessed by application calls in the same group which allows more dynamic behavior. For example an application can be created via Inner Transaction then funded immediately in the same transaction group since we have access to the created application id and address.   

# Smart Contract Storage 
Smart Contracts have three different types of storage: local storage, global storage, and box storage. 

> Global state and boxes are associated with the app itself, whereas local state is associated with each user account that opts into the application. Global and local storage are Key/Value pairs that are limited to 128 bytes per pair. Boxes are keyed storage segments up to 32kb of data per box.

Reviewer: ```suggestion
Global state and boxes are associated with the app itself, whereas local state is associated with each account that opts into the application. Global and local storage are Key/Value pairs that are limited to 128 bytes per pair. Boxes are keyed storage segments up to 32kb of data per box.
```

Each storage option’s properties are described below. 

**Global state for app a:**

* Allocation: 

>     * Can include between 0 and 64 key/value pairs and a total of 8K of memory to share among them.

Reviewer: ```suggestion
    * Can include between 0 and 64 key/value pairs for a total of 8K of memory to share among them.
```

>     * The amount of global storage is allocated in k/v units, and determined at contract creation. This cannot be edited later. 

Reviewer: ```suggestion
    * The amount of global storage is allocated in k/v units, and determined at contract creation. This schema is immutable after creation. 
```

    * The contract creator address is responsible for funding the global storage (by an increase to their minimum balance requirement, see below). 
* Reading: 
    * Can be read by any app call that has specified app a’s ID in its foreign apps array.
    * Can be read on-chain using the k/v pairs defined (from off-chain, can be read using goal or APIs + SDKs). 
* Writing: 
    * Can only be written by app a. 
* Deletion: 
    * Is deleted when app a is deleted. Cannot otherwise be deallocated (though of course the contents can be cleared by app a, but this does not change the minimum balance requirement). 

**Local state for account x for app a:** 

* Allocation: 
    * Is allocated when account x opts in to app a (submits a transaction to opt-in to app a).

>     * Can include between 0 and 16 key/value pairs and a total of 2KB of memory to share among them. 

Reviewer: ```suggestion
    * Can include between 0 and 16 key/value pairs for a total of 2KB of memory to share among them. 
```

    * The amount of local storage is allocated in k/v units, and determined at contract creation. This cannot be edited later. 
    * The opted-in user address is responsible for funding the local storage (by an increase to their minimum balance). 
* Reading: 
    * Can be read by any app call that has app x in its foreign apps array and account x in its foreign accounts array. 
    * Can be read on-chain using the k/v pairs defined (from off-chain, can be read using goal and the SDKs). 
* Writing: 
    * Is editable only by app a, but is delete-able by app a or the user x (using a ClearState call, see below). 
* Deletion: 

>     * Deleting an app does not affect its local storage. 

Reviewer: ??

Reviewer: wouldnt the local storage disappear? the min balance requirement would remain until the account opted out but id thought the local storage values would be inaccessible

    * _Clear state_. Every Smart Contract on Algorand has two programs: the _approval_ and the _clear state_ program. An account holder can clear their local state for an app at any time (deleting their data and freeing up their locked minimum balance). The purpose of the clear state program is to allow the app to handle the clearing of that local state gracefully. 
    * Account x can request to clear its local state using a [close out transaction](https://developer.algorand.org/docs/get-details/transactions/#application-close-out-transaction). 
    * Account x can clear its local state for app a using a [clear state transaction](https://developer.algorand.org/docs/get-details/transactions/#application-clear-state-transaction), which will always succeed, even after app a is deleted. 

**Boxes for app a:** 

* Allocation: 
    * App a can allocate as many boxes as it needs, when it needs them.

>     * App a allocates a box using the box_create opcode in its TEAL program, specifying the name and the size of the box being allocated. 

Reviewer: ```suggestion
    * App a allocates a box using the `box_create` opcode in its TEAL program, specifying the name and the size of the box being allocated. 
```

        * Boxes can be any size from 0 to 32K bytes. 
        * Box names must be at least 1 byte, at most 64 bytes, and must be unique within app a. 
    * The app account(the smart contract) is responsible for funding the box storage (with an increase to its minimum balance requirement, see below for details). 

>     * A box name must be referenced in the boxes array of the app call to be allocated. 

Reviewer: ```suggestion
    * A box name and app id must be referenced in the boxes array of the app call to be allocated. 
```

* Reading: 
    * App a is the only app that can read the contents of its boxes on-chain. This on-chain privacy is unique to box storage. Recall that everything can be read by anybody from off-chain using the algod or indexer APIs. 
    * To read box b from app a, the app call must include b in its boxes array. 
    * Read budget: Each box reference in the boxes array allows an app call to access 1K bytes of box state - 1K of “box read budget”. To read a box larger than 1K, multiple box references must be put in the boxes arrays. 
        * The box read budget is shared across the transaction group. 
        * The total box read budget must be larger than the sum of the sizes of all the individual boxes referenced (it is not possible to use this read budget for a part of a box - the whole box is read in).
    * Box data is unstructured. This is unique to box storage. 
    * A box is referenced by including its app ID and box name. 
* Writing: 
    * App a is the only app that can write the contents of its boxes.

>     * Exactly analogous to reading, each box ref in the boxes array allows an app call to write 1K bytes of box state - 1K of “box write budget”. 

Reviewer: I think we shouldn't use box read/box write budget and instead try to use `box io budget` or something so folks dont think there are 2 diff budgets

```suggestion
    * As with reading, each box ref in the boxes array allows an app call to write 1kb of box state - 1kb of “box write budget”. 
```

* Deletion: 
    * App a is the only app that can delete its boxes. 

>     * If an app is deleted, its boxes are not deleted. (the correct cleanup design is to look up the boxes from off-chain and call the app to delete all its boxes before deleting the app itself). 

Reviewer: aren't they deleted? the min balance would be locked if there are boxes outstanding but i thought the boxes themselves would also be gone

Reviewer: more a clarification, are the boxes indeed deleted? or do they remain possible to query from the algod api?

| Other App Global   |                  | `app_global_get_ex` |                  | `app_global_get_ex` |
| Other App Local    |                  | `app_local_get_ex`  |                  | `app_local_get_ex`  |

# Box Details
Boxes are useful in many scenarios:

* Applications that need larger or unbound contract storage.

> * Applications that want to store data per user, but do not wish to require users to opt-in to the contract.

Reviewer: ```suggestion
* Applications that want to store data per user, but do not wish to require users to opt-in to the contract or need the account data to persist even after the user closes or clears out of the application.
```

* Applications that require storing arbitrary maps or hash tables. 
  
The following sections cover the details of manipulating boxes within a smart contract. 

## Box Array 
The box array is an array of pairs: the first element of each pair is an integer specifying the index into the foreign application array, and the second element is the key name of the box to be accessed.

Each entry in the box array allows access to only 1kb of data. For example, if a box is sized to 4kb, the transaction must use four entries in this array. To claim an allotted entry a corresponding app Id and box name need to be added to the box ref array. If you need more than the 1kb associated with that specific box name, you can either specify the box ref entry more than once or, preferably, add “empty” box refs `[0,””]` into the array. If you specify 0 as the app Id the box ref is for the application being called. 

For example, suppose the contract needs to read “BoxA” which is 1.5kb, and “Box B” which is 2.5kb, this would require four entries in the box ref array and would look something like:

```py
boxes=[[0, "BoxA"],[0,"BoxB"], [0,""],[0,""]]
``` 

> The box reference budget is based on the sizes of the boxes accessed, not the amount of data read or written. For example, if a contract accesses “Box A” with a size of 2kb and “Box B” with a size of 10 bytes, this requires both boxes be in the box reference array and one additional reference, which should be an “empty” box reference. 

Reviewer: ```suggestion
The required box I/O budget is based on the sizes of the boxes accessed, not the amount of data read or written. For example, if a contract accesses “Box A” with a size of 2kb and “Box B” with a size of 10 bytes, this requires both boxes be in the box reference array and one additional reference ( ceil((2kb + 10b) / 1kb), which can be an “empty” box reference. 
```

Transaction 0: [0,”Box A”],[0,””],[0,””],[0,””],[0,””],[0,””],[0,””],[0,””]
Transaction 1: [0,””],[0,””],[0,””],[0,””],[0,””],[0,””],[0,””],[0,””]

Box refs can be added to the boxes array using `goal` or any of the SDKs.

=== "Goal"
    ```goal
    goal app method --app-id=53 --method="add_member2()void" --box="53,str:BoxA" --from=CONP4XZSXVZYA7PGYH7426OCAROGQPBTWBUD2334KPEAZIHY7ZRR653AFY
    ```

=== "Python SDK"
    ```py
    # Algorand Python SDK
    # Using ATC
    atc = AtomicTransactionComposer()
    atc.add_method_call(app_id, my_method, addr, sp, signer, method_args=[1,5], boxes=[[app_id, “key”]],
    )
    ```

=== "Beaker"
    ```py
    #Beaker framework
    result = app_client.call(
        Myapp.my_method,
    boxes=[[app_client.app_id, "key"]],
    )
    ```
## Creating a Box
The AVM supports two opcodes `box_create` and `box_put` that can be used to create a box. 

> The `box_create` opcode takes two parameters, the name and the size in bytes for the created box. The `box_put` opcode takes two parameters as well. The first parameter is the name and the second is a byte array to write. Because the AVM limits byte arrays to 4,096, `box_put` can only be used for boxes with length <= 4,096.

Reviewer: We should try to be consistent about specifying byte sizes (i.e. 4096 if we mean 4kb)

```suggestion
The `box_create` opcode takes two parameters, the name and the size in bytes for the created box. The `box_put` opcode takes two parameters as well. The first parameter is the name and the second is a byte array to write. Because the AVM limits any element on the stack to 4kb (#TODO link to stack parameter), `box_put` can only be used for boxes with length <= 4kb.
```

Reviewer: also, need to add the link to the parameter table here

!!!info
    When creating a box, the key name to be created must be in the box ref array.

## Writing to a Box  
The AVM provides two opcodes, `box_put` and `box_replace`,  to write data to a box. The `box_put` opcode is described in the previous section. The `box_replace` opcode takes three parameters, the key name, the starting location and replacement bytes.

=== "TEAL"
    ```teal
    byte “MyKey” 
    int 10
    byte “best”
    box_replace
    ```
=== "PyTeal"
    ```py
    #Beaker Framework
        @external
        def replace_string(self, ky: abi.String, start: abi.Uint64, replacement: abi.String, *, output: abi.String):
            return Seq(
                App.box_replace(ky.get(), start.get(), replacement.get()),
                boxstr :=  App.box_get(ky.get()),
                Assert( boxstr.hasValue()),
                output.set(boxstr.value()),
            ) 

    ```

When using `box_replace`, the box size can not increase. This means if the replacement bytes, when added to the start byte location, exceed the upper bounds of the box, the call will fail. 

## Reading from a Box

> The AVM provides two opcodes for reading the contents of a box, `box_get` and `box_extract`. The `box_get` opcode takes one parameter which is the key name for the box. It reads the entire contents of a box. The `box_get` opcode returns two values. The top-of-stack is an integer that has the value of 1 or 0. A value of 1 means that the box was found and read. A value of 0 means that the box was not found. The next stack element contains the bytes read if the box exists, else it contains an empty byte array. `box_get` fails if the box length exceeds 4,096.

Reviewer: ```suggestion
The AVM provides two opcodes for reading the contents of a box, `box_get` and `box_extract`. The `box_get` opcode takes one parameter which is the key name for the box. It reads the entire contents of a box. The `box_get` opcode returns two values. The top-of-stack is an integer that has the value of 1 or 0. A value of 1 means that the box was found and read. A value of 0 means that the box was not found. The next stack element contains the bytes read if the box exists, else it contains an empty byte array. `box_get` fails if the box length exceeds 4kb.
```

    ```
    100,000 + 28,500 = 128,500
    ```

# Minimum Balance Requirement For Boxes

> Boxes are created by a smart contract and raise the minimum balance requirement(MBR) that must be in the contract's ledger balance. This means that a contract that intends to use boxes, must be funded beforehand.

Reviewer: ```suggestion
Boxes are created by a smart contract and raise the minimum balance requirement (MBR) that must be in the contract's ledger balance. This means that a contract that intends to use boxes, must be funded beforehand.
```
